File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.cpp
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        vector<int> b(k, INT_MIN); // buy states
        vector<int> s(k, 0); // sell states
        int i;
        for (int p: prices) {
            i = 0;
            while (i < k) {
                if (i == 0) b[i] = max(b[i], -p);
                else b[i] = max(b[i], s[i-1] - p);
                s[i] = max(s[i], b[i] + p);
                i++;
            }
        }
        return s[k-1];
    }
};

File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        minCosts = [1001 for _ in range(k)]
        maxProfits = [0 for _ in range(k)]
        for val in prices:
            for i in range(k):
                minCosts[i] = min(minCosts[i], val - (0 if i == 0 else maxProfits[i-1]))
                maxProfits[i] = max(maxProfits[i], val - minCosts[i])
        return maxProfits[-1]